qtfred/src/mission/dialogs/CampaignEditorDialogModel.cpp

```diff
@@ -0,0 +1,801 @@
+
+	for (auto shp_idx_ptr : initialShips.getCheckedData()) {
+		Assertion(shp_idx_ptr, "NULL ship class index in initial ships");
+		Assertion(*shp_idx_ptr < MAX_SHIP_CLASSES, "Illegal ship class index in initial ships: %ld", *shp_idx_ptr);
+		Campaign.ships_allowed[*shp_idx_ptr] = true;
+	}
+
+	for (auto wep_idx_ptr : initialWeapons.getCheckedData()) {
+		Assertion(wep_idx_ptr, "NULL weapon class index in initial weapons");
+		Assertion(*wep_idx_ptr < MAX_WEAPON_TYPES, "Illegal weapon class index in initial weapons: %ld", *wep_idx_ptr);
+		Campaign.weapons_allowed[*wep_idx_ptr] = true;
+	}
+
+	if (firstMission.length() > 0) {
+		static const QString PAST_BRANCHES{};
+		SCP_unordered_set<const CampaignMissionData*> unsaved{missionData.getCheckedData()};
+		SCP_queue<const QString*> saveNext{};
+		int i=0, lvl=0, pos=0;
+
+		saveNext.push(&firstMission);
+		do {
+			// traversal
+			if (saveNext.empty())
+				saveNext.push(&(*unsaved.cbegin())->filename);
+			while (! saveNext.empty()) {
+				const QString *mnName = saveNext.front();
+				saveNext.pop();
+
+				if (mnName == &PAST_BRANCHES) {
+					if (pos > 0) {
+						++lvl;
+						pos = 0;}
+					continue;
+				}
+
+				auto it = std::find_if(unsaved.cbegin(), unsaved.cend(),
+									   [&](const CampaignMissionData *mn_ptr) {
+					 return mn_ptr->filename == *mnName;	});
+				const CampaignMissionData *mn;
+				if (it != unsaved.cend()) {
+					mn = *it;
+					unsaved.erase(it);
+				}
+				else {
+					continue;
+				}
+
+				for (const auto &br : mn->branches)
+					saveNext.push(&br.next);
+				saveNext.push(&PAST_BRANCHES);
+
+				// saving
+				cmission &cm = Campaign.missions[i++];
+
+				cm.name = vm_strdup(qPrintable(*mnName));
+				strncpy(cm.briefing_cutscene, qPrintable(mn->briefingCutscene), NAME_LENGTH);
+				cm.main_hall = mn->mainhall.toStdString();
+
+				//Bastion flag unsupported in missionLoad
+				cm.flags = 0;
+				cm.debrief_persona_index = static_cast<ubyte>(mn->debriefingPersona.toUShort());
+
+				using BranchType = CampaignBranchData::BranchType;
+				cm.formula =
+						alloc_sexp("cond", SEXP_ATOM, SEXP_ATOM_OPERATOR, -1, -1);
+				int *cond_arms_ptr = &Sexp_nodes[cm.formula].rest;
+				bool flag_last_branch = false;
+				for (const auto &br : mn->branches) {
+					if (br.type == BranchType::INVALID) {
+						QMessageBox::warning(nullptr, "Can't save", "Invalid branch: From " + mn->filename + " to " + br.next);
+						return false;
+					}
+					if (br.type == BranchType::NEXT_NOT_FOUND)
+						QMessageBox::warning(nullptr, "Potential Campaign Bug", "Saving branch to unknown mission: " + br.next);
+					if (! br.loop) { //build formula from nonloop branches
+						if (flag_last_branch)
+							QMessageBox::warning(nullptr, "Potential Campaign Bug", "Branch is unreachable due to previous \"true\" condition: " + br.next);
+						if (br.sexp == Locked_sexp_true)
+							flag_last_branch = true;
+						int nextsexp = (br.type != BranchType::END) ?
```

Comment:
This is a very loaded, multi-line usage of the trinary operator.  Use an if-else statement instead.